jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/paths/FileSystemPool.java

```diff
@@ -0,0 +1,114 @@
+//
+// ========================================================================
+// Copyright (c) 1995-2022 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.util.paths;
+
+import java.io.IOException;
+import java.net.URI;
+import java.nio.file.FileSystem;
+import java.nio.file.FileSystems;
+import java.nio.file.Path;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * A pool of {@code zipfs} based {@code java.nio.file.FileSystem} objects to allow reference
+ * counting of open FileSystem and proper closure when the last reference count is decremented.
+ */
+public class FileSystemPool
+{
+    private static final FileSystemPool INSTANCE = new FileSystemPool();
+
+    public static FileSystem acquire(Path jarfile) throws IOException
+    {
+        return INSTANCE.acquireZipFs(jarfile, "runtime");
+    }
+
+    public static boolean release(Path jarfile) throws IOException
+    {
+        return INSTANCE.releaseZipFs(jarfile);
+    }
+
+    protected static Map<Path, FileSystemRefCount> getCache()
+    {
+        return INSTANCE.cache;
+    }
+
+    public static class FileSystemRefCount
+    {
+        private Path rootPath;
+        private FileSystem fileSystem;
+        private final AtomicInteger count = new AtomicInteger(0);
+    }
+
+    private final Map<Path, FileSystemRefCount> cache = new ConcurrentHashMap<>();
+
+    public FileSystem acquireZipFs(Path jarFile, String multiRelease) throws IOException
+    {
+        FileSystemRefCount refCount = cache.get(jarFile);
+        if (refCount == null)
+        {
+            Map<String, String> env = new HashMap<>();
+            env.put("multi-release", multiRelease);
+
+            URI uri = URI.create("jar:" + jarFile.toUri().toASCIIString());
+            refCount = new FileSystemRefCount();
+            refCount.fileSystem = FileSystems.newFileSystem(uri, env);
+            refCount.rootPath = refCount.fileSystem.getPath("/");
+            cache.put(jarFile, refCount);
+        }
+
+        refCount.count.incrementAndGet();
+        return refCount.fileSystem;
+    }
```

Comment:
This needs to be atomic.  I know it is very unlikely to be hit concurrently, but it is possible.  So a null check then add is not sufficient as two threads can add at the same time.
Probably just needs a bit lock as any of the IfAbsent  style method might still create a new FS that has to be discarded.
Hmmm maybe @lorban has a pool that can be used?

Reply:
If the pool class can support this use case, I'm all for moving to a standardized pooling layer.

Reply:
The easiest fix would be to use `ConcurrentMap.computeIfAbsent()` to create the entry. The `Pool` class is not a key-value store so it does bit the bill of this use-case.

Reply:
I'm not sure computeIfAbsent is good enough as it can end up creating an entry which is ultimately not used.  So in this case it would create a zip filesystem that loses the race to be added and then has to be disposed of.
I think a big fat lock is probably best here.  Not on critical path.

Reply:
I think you're misunderstanding `computeIfAbsent`: the lambda is only called when the entry is null _and_ you won the race. If the entry isn't null, it's simply returned as a `get` would. I think it's safe here, but I'm not against using a lock instead.